iina/MPVController.swift

```diff
@@ -1130,118 +1136,148 @@ not applying FFmpeg 9599 workaround
+        logPropertyValueError(MPVProperty.videoParamsRotate, property.format)
+        break
       }
+      DispatchQueue.main.async { self.player.mainWindow.rotation = rotation }
 
     case MPVProperty.videoParamsPrimaries:
       fallthrough;
 
     case MPVProperty.videoParamsGamma:
-      player.refreshEdrMode()
+      if #available(macOS 10.15, *) {
+        DispatchQueue.main.async { self.player.refreshEdrMode() }
+      }
 
     case MPVOption.TrackSelection.vid:
-      player.vidChanged()
+      DispatchQueue.main.async { self.player.vidChanged() }
 
     case MPVOption.TrackSelection.aid:
-      player.aidChanged()
+      DispatchQueue.main.async { self.player.aidChanged() }
 
     case MPVOption.TrackSelection.sid:
-      player.sidChanged()
+      DispatchQueue.main.async { self.player.sidChanged() }
 
     case MPVOption.Subtitles.secondarySid:
-      player.secondarySidChanged()
+      DispatchQueue.main.async { self.player.secondarySidChanged() }
 
     case MPVOption.PlaybackControl.pause:
-      if let paused = UnsafePointer<Bool>(OpaquePointer(property.data))?.pointee {
+      guard let paused = UnsafePointer<Bool>(OpaquePointer(property.data))?.pointee else {
+        logPropertyValueError(MPVOption.PlaybackControl.pause, property.format)
+        break
+      }
+      DispatchQueue.main.async { [self] in
         if player.info.isPaused != paused {
           player.sendOSD(paused ? .pause : .resume)
-          DispatchQueue.main.sync {
-            player.info.isPaused = paused
-            player.refreshSyncUITimer()
-            if paused {
-              player.mainWindow.videoView.displayIdle()
-            } else {
-              player.mainWindow.videoView.displayActive()
-            }
+          player.info.isPaused = paused
+          player.refreshSyncUITimer()
+          // Follow energy efficiency best practices and ensure IINA is absolutely idle when the
+          // video is paused to avoid wasting energy with needless processing. If paused shutdown
+          // the timer that synchronizes the UI and the high priority display link thread.
+          if paused {
+            player.mainWindow.videoView.displayIdle()
+          } else {
+            player.mainWindow.videoView.displayActive()
           }
         }
         if player.mainWindow.loaded && Preference.bool(for: .alwaysFloatOnTop) {
-          DispatchQueue.main.async {
-            self.player.mainWindow.setWindowFloatingOnTop(!paused)
-          }
+          player.mainWindow.setWindowFloatingOnTop(!paused)
         }
+        player.syncUI(.playButton)
       }
-      player.syncUI(.playButton)
 
     case MPVProperty.chapter:
-      player.info.chapter = Int(getInt(MPVProperty.chapter))
-      player.syncUI(.time)
-      player.syncUI(.chapterList)
-      player.postNotification(.iinaMediaTitleChanged)
+      DispatchQueue.main.async { [self] in
+        player.info.chapter = Int(getInt(MPVProperty.chapter))
```

Comment:
I'm confused. Why can't `getInt` be called on the main thread?

Reply:
Because it's `mpv_get_property()`, and we want it to be on the `com.colliderli.iina.controller` thread. If it's on the main thread, then for example, `mpv_destroy()` is on the `com.colliderli.iina.controller` thread, and it will cause another potential crash due to race condition when closing the window.

https://github.com/iina/iina/blob/1ebe0f459627c8e10639bd65276e29bd57629fdd/iina/MPVController.swift#L970-L975

Reply:
Will fix, but we do call methods like `getInt` outside of `MPVController` from code I'm sure is running on the main thread. I will think on how to better coordinate this. I have seen `mpv` emit some events during shutdown that it really shouldn't be emitting. I've been worried that we let `q` through to `mpv` causing `mpv` to shutdown "behind" IINA's back although I've not been able to trigger a crash in testing that. The `gets` are a problem because they do not return optional values. If they did it would be easy to detect shutdown and return `nil`.

Reply:
Yes, but these ones are perhaps much easier to trigger because it's in `handleEvent`. We can't say exactly that "mpv is emitting events during shutdown", because this block is queued on the main thread and executed later than the event, after some unknown delay. If mpv shuts down right after emitting the event, there will be a crash.

> The gets are a problem because they do not return optional values. If they did it would be easy to detect shutdown and return nil.

If we are going to add the check, I think we can just return 0. If mpv is already terminated then the value makes little sense.

Reply:
Brain wasn't remembering this PR last night, it has been a while since I coded it. I remember thinking having the get methods return some random value seemed a bit wrong. So I decided it would be better to coordinate shutdown at a higher level. I have been slowly refactoring MPVController to call "listener" methods in PlayerCore. Those methods have the ability to check if shutdown is in progress and if so, drop the event. I need to refactor the code you pointed out into additional "listener" methods in PlayerCore that check the state and avoid calling `getInt`.

I'm used to a layered design where the bottom layer classes know very little about the higher level code. I was expecting we'd have a class that was focused solely on providing a Swift interface for the mpv client. Since we did not have such a class I did not consider implementing the interesting suggestion from @CarterLi in PR #4367 of using mpv to display previews of JPEG XL screenshots. In macOS Sonoma AppKit renders the preview of JPEG XL screenshots, so going forward mpv is not needed for that. However we still at some point need support for HDR screenshots. One way to do that would be to move away from our proprietary thumbnail file format and store the thumbnails as a MP4 and use mpv to display them. A discussion for another day. Anyway, just an example of how a more isolated mpv client could have been used.

Reply:
I think I have addressed the comments. I added a comment about the main thread requirement. I moved the code that was calling back into mpv to PlayerCore with checks for shutting down.